WIP: Demo: Offer Tokenizing Improvements for Highlighting

msftrncs opened the pull request:
In function name and command name parsing, insure that an expandable string
does not exist as these are not to be expanded.
In command name parsing, insure that redirect tokens that will be accepted as
commands are flagged as such.
In command argument processing, arguments that originate as 'Operator' tokens have their TokenFlags reverted when switched from an 'operator' TokenKind to a 'Generic' TokenKind.
In command name processing, treat `DashDash` token as a command name and do not set in motion the 'no more parameters' mode.
In command argument processing, correct the 'no more parameters' mode so additional 'parameter' tokens are not marked as command names.
In command name processing, treat `--%` (verbatim argument operator) as a command and do not treat next argument verbatim.
In function definitions, correct function name token of `--` to be generic, and fix assumptions that generic tokens are string tokens.

Reference #10275, #10347, #10348

to do
- [ ] Find tests for `--`, add tests to make sure `--` after a command of `--` still works.
- [ ] Find tests for `--%`, add tests to make sure `--%` after a command of `--%` still works.
- [ ] Find tests for redirection operators, add tests to make sure redirection operators as commands do not interfere with redirection operators later in the command statement.

# PR Summary

Adjust token polishing to benefit token based highlighting.

## PR Context

The syntax highlighting offered by PSReadLine has always led me through some confusion and I am sure every once in a while it confuses some others as well as to what is really accepted as syntax.  However the issue is not really with PSReadLine but with the tokens that PSReadLine uses for highlighting, rather than using the AST.

It would appear that some house keeping tasks are not performed on the token collection after the AST is produced from them.  Some places clean up (set indicator flags) and some places do not.  The most notable item is that tokens that ultimately become `StringExpandableToken` but when processed in to AST are treated as just `StringToken` remain in that form, which then incorrectly informs any downstream applications.

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

msftrncs committed:
Offer Tokenizing Improvements for Highlighting

In function name and command name parsing, insure that an expandable string
does not exist as these are not to be expanded.
In command name, insure that redirect tokens that will be accepted as
commands are flagged as such.

Reference #10275

JamesWTruher reviewed (changes requested):
definitely going to want to see a way to test the changes here

msftrncs commented:
@JamesWTruher, I think tests could be added to `test\powershell\language\parser\parsing.tests.ps1`, some `ParseInput` expressions could be constructed to test the tokenizer results of a few sample commands.  I am referencing the ternary operator parsing tests that @daxian-dbw added.  In some ways this is similar to how PSReadLine performs tests of its highlighting, but without simulating the input and output steps.

- `ParseInput` a sample command,
- verify select tokens are as expected (either by `TokenKind` or `TokenFlag`).

I'll work on getting this done.

msftrncs commented:
I've included another commit to demonstrate all the potential highlighting improvements I think should be made within parsing/tokenizing.  However, this also includes two changes to behavior, for `--` and `--%` when used as command names (via invoke operators).  The referenced issues are listed above the PR summary.

lzybkr commented on src/System.Management.Automation/engine/parser/Parser.cs:
```diff
@@ -5284,6 +5284,13 @@ private StatementAst FunctionDeclarationRule(Token functionToken)
             functionNameToken.TokenFlags &= ~TokenFlags.Keyword;
 
+            // A function name is not an expandable token, so remove any nested tokens
+            var expandableToken = functionNameToken as StringExpandableToken;
```
Pattern matching makes this a little easier to read:

```c#
if (functionNameToken is StringExpandableToken expandableToken)
{
    expandableToken.NestedTokens = null;
}
```

msftrncs replied:
ahhh… that's what that suggestion message means?  I had copied that from a similar part in `GetCommandArgument`.

lzybkr replied:
I'm not sure what suggestion you're referring to, but pattern matching was not available in C# when the code you copied was written.

msftrncs replied:
I've addressed this code in revised commit, but have a ways to go on tests.

msftrncs replied:
VS Code suggests all over 'Use Pattern Matching'.  The area of the code I copied was not one such area though, probably because its too complex for the pattern matching syntax.

lzybkr reviewed (commented):
Changes look reasonable, but it's important to add tests that validate no semantic changes, e.g passing `--` after `--` or using a redirection operator as a command name.

msftrncs commented:
I've revised the most recent commit to address another `--` highlighting issue, as a function name in a function definition.   For this I used `Token.SetAsCommandArgument`.  This method should probably be more aptly named `SetAsGenericTokenKind`.   However, there was further code that assumed that a `TokenKind.Generic` token would be a `StringToken`, so I had to make some adjustments.

msftrncs committed:
Address `--` and `--%` highlight issues

Correct tokenizing output so that `--` when used as a command name does not
instigate the 'stop processing parameters' function and is correctly
identified as a command.  Also correct further parameter tokens from being
incorrectly marked as commands while in that mode.
Correct tokenizing treatment of `--%` when it is a command name so as not
server as the VerbatimArgumentMode, and also to be indentified as a
command name token.
Address tokenizing of `--` (and others) as a function name, using
Token.SetIsCommandArgument.
This also addresses token flags being left set when operator tokens are
converted to generic tokens in Token.SetIsCommandArgument.
Also addressed assumptions that TokenKind.Generic is always StringToken.
